### ISSUE TITLE
[Bug]: Using OneSignal.login() does not set externalId in Subscription Records

### ISSUE THREAD

**Issue by sebastinto** [OneSignal/OneSignal-Android-SDK#2043]
### What happened?

Calling `OneSignal.login()` does not seem to be updating Subscription Records i.e. the External ID column values are unset. `OneSignal.User.externalId` is correctly set on device.

I tried on a fresh install, multiple device, calling `OneSignal.logout()` before `OneSignal.login()` to no avail. I'm wondering if there's anything I may be missing.

Tags are set correctly.

### Steps to reproduce?

```Markdown
1. Initialize SDK
2. Call OneSignal.login() with a value (e.g. "1234")
3. Observe Subscription Records External ID does not update
```

### What did you expect to happen?

External ID updates in Subscription Records.

### OneSignal Android SDK version

Release 5.1.7

### Android version

13

### Specific Android models

```Markdown
* Pixel 8 Pro
* Samsung S20 5G
```

### Relevant log output

[os_log.txt](https://github.com/OneSignal/OneSignal-Android-SDK/files/14875464/os_log.txt)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by sebastinto** [comment]
Thanks @jkasten2 !

> 1. Where are you calling `OneSignal.initWithContext` and `OneSignal.login` in your code?

`OneSignal.initWithContext` is called in the `onCreate()` function of the main `Application` class. `OneSignal.login` is called after a user successfully authenticates via our API.

> 2. "Observe Subscription Records External ID does not update" where are you observing this? On the OneSignal dashboard?

Correct!

> 3. Does the issue happen if you simply call `OneSignal.initWithContext` then  `OneSignal.login`?
>    * Or does the issue only happen if you call logout()?

Removing `logout()` fixes the issue.

I have a few follow-up questions if you don't mind:

1. Is this expected behavior? Why would calling `logout()` prevent subsequent `login()` calls?
2. It seems like calling `login()` with a different value is enough to override external id. Is this the recommended implementation? 
3. If it is, what would be the use case for calling `logout()`? Explicitly de-identifying a user? How would you re-identify a user after calling `logout()`?

**Comment by jkasten2** [comment]
> Removing logout() fixes the issue.

Good to hear! However that is bug as `logout()` shouldn't cause an issue like this. Could you explain step-by-step what the order of timing of the login() and logout() calls that cause the issue? Include any delays or if they are called back-to-back.

> 1. Is this expected behavior? Why would calling logout() prevent subsequent login() calls?

It seems to be a bug, the repo steps noted above will be helpful to us.

>2. It seems like calling login() with a different value is enough to override external id. Is this the recommended implementation?

Yes, this usage is something we design for, if you call login with a different id it simply switches users.

> If it is, what would be the use case for calling logout()? Explicitly de-identifying a user? How would you re-identify a user after calling logout()?

The only reason to call logout is if you have that functionally in your app. What OneSignal does is create a new anonymous User and move the push subscription to it. This way the subscription has been dissociated from the user that was logged in.

**Comment by sebastinto** [comment]
Dig a little bit more digging and it seems like calling `logout()` before `login()` may not be the issue after all. Sorry about that.

External IDs do update in the OneSignalDashboard if they **do not already exist** even when `logout()` is called before `login()` (which may happen back-to-back, a short second apart to answer your question).

However calling `login()` and passing the value of an **existing** External ID (i.e. one that can already be found in the OneSignal Dashboard), results in a new entry that has a blank External ID value.

It might be worth mentioning that we recently replaced the Service Account.json with one from a different Firebase project, but I just rolled back to the previous version of the OneSignal SDK that we were using (4.8.4) and existing External IDs do update as expected in the OneSignal Dashboard.

> Yes, this usage is something we design for, if you call login with a different id it simply switches users.

Nice, good to know it is an option.

> The only reason to call logout is if you have that functionally in your app. What OneSignal does is create a new anonymous User and move the push subscription to it. This way the subscription has been dissociated from the user that was logged in.

That makes sense. We do have that functionality in the app. Users can explicitly log out / delete local data / analytics id.

**Comment by jkasten2** [comment]
Thanks for all the details, I have undercovered some cases we are not handling correctly and we will improve on this in the next release.

**Comment by sebastinto** [comment]
> Thanks for all the details, I have undercovered some cases we are not handling correctly and we will improve on this in the next release.

Interesting! I'll definitely keep an eye out for an update, thanks!

**Comment by jkasten2** [comment]
@sebastinto we released [5.1.8](https://github.com/OneSignal/OneSignal-Android-SDK/releases/tag/5.1.8) which should address this issue. Thanks for the details you provided here!

Let us know if the issue is completely gone after this update or not.

**Comment by sebastinto** [comment]
> @sebastinto we released [5.1.8](https://github.com/OneSignal/OneSignal-Android-SDK/releases/tag/5.1.8) which should address this issue. Thanks for the details you provided here!
> 
> Let us know if the issue is completely gone after this update or not.

Fix confirmed! Thanks!
